apps/coordinator/src/selectors/transaction.ts

```
@@ -0,0 +1,244 @@
+          } else {
+            console.warn(
+              `❌ Invalid signature for input ${inputIndex}, pubkey ${pubkey}`,
+            );
+            isCompleteSet = false;
+            break;
+          }
+        } catch (error) {
+          console.warn(
+            `❌ Error validating signature for input ${inputIndex}:`,
+            error.message,
+          );
+          isCompleteSet = false;
+          break;
+        }
+      } else {
+        // This pubkey didn't sign this input, so it's not a complete set
+        isCompleteSet = false;
+        break;
+      }
+    }
+
+    // Only add complete signature sets (where one signer signed ALL inputs)
+    if (isCompleteSet && signatureSet.length === inputs.length) {
+      signatureSets.push({
+        signatures: signatureSet,
+        publicKeys: publicKeySet,
+        signerPubkey: pubkey,
+      });
+    }
+  }
+
+  return signatureSets;
+}
+
+/**
+ * Validates a signature for a specific input.
+ *
+ * *Note* not a selector a plain util function
+ *
+ * @param signature - Hex-encoded signature
+ * @param inputIndex - Index of the input being signed
+ * @param originalPsbt - The original PSBT object we already have
+ * @param inputs - All transaction inputs (for amountSats)
+ * @returns The public key if valid, false if invalid
+ */
+function validateSignatureForInput(
+  signature: Buffer,
+  inputIndex: number,
+  originalPsbt: Psbt,
+  inputs: Input[],
+): string | false {
+  try {
+    //  Handle empty signatures
+    if (!signature || signature.length === 0) {
+      return false;
+    }
+
+    const validatedPubkey = validateMultisigPsbtSignature(
+      originalPsbt.toBase64(),
+      inputIndex,
+      signature,
+      inputs[inputIndex].amountSats,
+    );
+
+    // Handle the case where validateMultisigPsbtSignature returns boolean true
+    // Convert true to false since we only want string (pubkey) or false
+    if (validatedPubkey === true) {
```

I don't understand what's going on here? why would it be true but then we want to say it's false? Let's just throw an error if it's false. Then we don't even need to check for booleans.

---

Ahh, right — I remember now. I added that as a quick workaround to satisfy the type checker. Since `validateMultisigPsbtSignature` only ever returns `false`, but its return type is defined as `boolean`, the type checker was shouting at me 😅

So yeah... I did this dumb hacky fix just to keep things passing at the time. I’ll clean it up properly now!